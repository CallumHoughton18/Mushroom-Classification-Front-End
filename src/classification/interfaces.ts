import {IFormElement} from "../shared/interfaces";

export interface IClassificationData {
    fieldName: string;
    options: Array<string>;
}

export interface IClassificationQuestion extends IFormElement, IClassificationData {}
