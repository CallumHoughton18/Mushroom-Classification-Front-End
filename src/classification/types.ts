import {FormContents, SelectOption} from "../shared/types";

export type APIGet<T> = {
    success: boolean;
    result: T;
};

export type ClassificationResponse = {
    poisonous: boolean;
};

export type ClassificationQueryData = {
    classificationData: FormContents;
};

export type FeatureDefinition = {
    name: string;
    options: SelectOption[];
};
