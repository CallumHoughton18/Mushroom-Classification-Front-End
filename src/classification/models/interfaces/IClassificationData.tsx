interface IClassificationData {
    fieldName: string;
    options: Array<string>;
}

export default IClassificationData;
