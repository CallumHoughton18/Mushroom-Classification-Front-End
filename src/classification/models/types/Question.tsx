type Question = {
    title: string;
    options: Array<string>;
};

export default Question;
