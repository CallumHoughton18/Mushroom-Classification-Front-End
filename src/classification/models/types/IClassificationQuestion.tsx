import IClassificationData from "../interfaces/IClassificationData";
import IFormElement from "../../../shared/interfaces/IFormElement";

interface IClassificationQuestion extends IFormElement, IClassificationData {}

export default IClassificationQuestion;
