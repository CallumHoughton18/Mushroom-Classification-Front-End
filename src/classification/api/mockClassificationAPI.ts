import {IClassificationAPI} from "../interfaces";
import {FeatureDefinition, APIGet} from "../types";

const mockClassificationAPI: IClassificationAPI = {
    getClassification: async (): Promise<APIGet<boolean>> => {
        await new Promise((resolve) => setTimeout(resolve, 400));
        return {
            success: true,
            result: true
        };
    },
    getClassificationFormDefinition: async (): Promise<APIGet<FeatureDefinition[]>> => {
        const mockDefs: FeatureDefinition[] = [];
        for (let index = 0; index < 3; index++) {
            const classificationQuestions: FeatureDefinition = {
                name: `classques-${index}`,
                options: []
            };
            mockDefs[index] = classificationQuestions;
        }
        return {success: true, result: mockDefs};
    }
};

export default mockClassificationAPI;
