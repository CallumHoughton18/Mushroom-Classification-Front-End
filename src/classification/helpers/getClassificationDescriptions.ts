import data from "../../assets/files/more-info.json";
import {ClassificationDescriptions} from "../types";

/**
 * @returns {ClassificationDescriptions} JSON stored in "more-info.json".
 */
const getClassificationDescriptions = (): ClassificationDescriptions => {
    return (data as unknown) as ClassificationDescriptions;
};

export default getClassificationDescriptions;
