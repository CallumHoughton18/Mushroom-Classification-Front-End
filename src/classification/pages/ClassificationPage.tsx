import React, {FunctionComponent} from "react";
import ClassificationForm from "../components/ClassificationForm";
import useAppNavigation from "../../navigation/hooks/useAppNavigation";
import {FormContents} from "../../shared/types";
import {ClassificationQueryData} from "../types";
import {IClassificationAPI} from "../interfaces";
import {useGetFormDefinition} from "../api/classificationAPIHooks";
import {LoadingState} from "../../shared/enums";
import Spinner from "../../shared/components/UI/Spinner";
import useMoreInfoModal from "../hooks/useMoreInfoModal";
import getClassificationDescriptions from "../helpers/getClassificationDescriptions";

type ClassificationPageProps = {
    classificationAPI: IClassificationAPI;
};

const ClassificationPage: FunctionComponent<ClassificationPageProps> = ({classificationAPI}) => {
    const classificationDescs = getClassificationDescriptions();

    const navManager = useAppNavigation();
    const [modal, setQuesForInfo, toggle] = useMoreInfoModal(classificationDescs);
    const [formQuestions, loading] = useGetFormDefinition(classificationAPI, () => {
        navManager.goToErrorPage({message: "Error fetching form definition"});
    });

    if (loading === LoadingState.LOADING) return <Spinner />;

    const navToClassificationResult = (formData: FormContents) => {
        const classificationData: ClassificationQueryData = {
            classificationData: formData
        };

        navManager.goToClassificationResultPage(classificationData);
    };

    return (
        loading === LoadingState.LOADED && (
            <div>
                {modal}
                <ClassificationForm
                    questions={formQuestions}
                    onSubmit={navToClassificationResult}
                    hasInfoPopup={(question) => {
                        return classificationDescs[question.fieldName] != null;
                    }}
                    viewInfoCallback={(question) => {
                        console.log(question);
                        setQuesForInfo(question);
                        toggle();
                    }}
                ></ClassificationForm>
            </div>
        )
    );
};

export default ClassificationPage;
