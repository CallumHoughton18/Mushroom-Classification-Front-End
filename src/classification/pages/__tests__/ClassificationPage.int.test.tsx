import {render, waitFor} from "@testing-library/react";
import React, {ReactNode} from "react";
import {mocked} from "ts-jest/utils";
import userEvent from "@testing-library/user-event";
import useAppNavigation from "../../../navigation/hooks/useAppNavigation";
import ClassificationPage from "../ClassificationPage";
import mockClassificationAPI from "../../api/mockClassificationAPI";
import {IClassificationAPI} from "../../interfaces";
import {APIGet, FeatureDefinition} from "../../types";
import {generateMockNav} from "../../../test_helpers/testHelpers";
import getClassificationDescriptions from "../../helpers/getClassificationDescriptions";
import {moreInfoRole} from "../../../shared/constants";

jest.mock("../../../navigation/hooks/useAppNavigation");
jest.mock("../../../shared/hooks/domHooks", () => {
    return {
        useGetElement: () => {
            return {};
        }
    };
});
jest.mock("../../helpers/getClassificationDescriptions");
jest.mock("react-dom", () => {
    const org = jest.requireActual("react-dom");
    return {
        ...org,
        createPortal: (node: ReactNode) => node
    };
});

describe("<Classificationpage /> functional tests", () => {
    const mockedUseAppNavigation = mocked(useAppNavigation, true);
    const mockedGetClassificationDescriptions = mocked(getClassificationDescriptions, true);
    const mockNav = generateMockNav();

    beforeAll(() => {
        mockedUseAppNavigation.mockImplementation(() => {
            return mockNav;
        });
        mockedGetClassificationDescriptions.mockImplementation(() => {
            return {
                "classques-1": {info: "information description goes here"},
                "classques-2": {info: "information description goes here", image: "test-image.png"}
            };
        });
    });

    beforeEach(() => {
        mockedUseAppNavigation.mockClear();
        mockedGetClassificationDescriptions.mockClear();
    });

    it("Should perform navigation on successful form submission", () => {
        const {getByRole} = render(
            <ClassificationPage classificationAPI={mockClassificationAPI} />
        );

        waitFor(() => {
            userEvent.click(getByRole("button"));
            expect(mockNav.goToClassificationResultPage).toHaveBeenCalledTimes(1);
        });
    });

    it("Should navigate to error page if getFormDef hook fails", () => {
        const classificationAPIUnSuccesseMock: IClassificationAPI = {
            getClassification: mockClassificationAPI.getClassification,
            getClassificationFormDefinition: () => {
                const fail: APIGet<FeatureDefinition[]> = {
                    success: false,
                    result: []
                };
                return Promise.resolve(fail);
            }
        };

        render(<ClassificationPage classificationAPI={classificationAPIUnSuccesseMock} />);
        waitFor(() => {
            expect(mockNav.goToErrorPage).toHaveBeenCalledTimes(1);
        });
    });

    it("Should display modal", () => {
        const page = render(<ClassificationPage classificationAPI={mockClassificationAPI} />);

        waitFor(() => {
            const moreInfoBtns = page.getAllByRole(moreInfoRole)[0];
            moreInfoBtns.click();
            expect(page).toMatchSnapshot();
        });
    });
});
