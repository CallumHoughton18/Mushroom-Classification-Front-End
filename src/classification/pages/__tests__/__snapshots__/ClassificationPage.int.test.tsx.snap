// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Classificationpage /> functional tests Should display modal 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="modal"
        >
          <section
            class="modal-window"
          >
            <div
              class="modal-header"
            >
              Classques 1
            </div>
            <div
              class="modal-body"
            >
              <div
                class="classification-info"
              >
                <p>
                  information description goes here
                </p>
              </div>
            </div>
            <div
              class="modal-footer"
            >
              <button>
                Close
              </button>
            </div>
          </section>
        </div>
        <form>
          <label
            for="classques-0"
          >
            Classques 0
             
          </label>
          <select
            id="classques-0"
            name="classques-0"
            required=""
            title="Classques 0"
          >
            <option
              disabled=""
              value=""
            >
              Select an option
            </option>
          </select>
          <label
            for="classques-1"
          >
            Classques 1
             
            <a
              role="more-info-button"
            >
              <i
                class="fa"
              />
            </a>
          </label>
          <select
            id="classques-1"
            name="classques-1"
            required=""
            title="Classques 1"
          >
            <option
              disabled=""
              value=""
            >
              Select an option
            </option>
          </select>
          <label
            for="classques-2"
          >
            Classques 2
             
            <a
              role="more-info-button"
            >
              <i
                class="fa"
              />
            </a>
          </label>
          <select
            id="classques-2"
            name="classques-2"
            required=""
            title="Classques 2"
          >
            <option
              disabled=""
              value=""
            >
              Select an option
            </option>
          </select>
          <input
            type="submit"
            value="Submit"
          />
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="modal"
      >
        <section
          class="modal-window"
        >
          <div
            class="modal-header"
          >
            Classques 1
          </div>
          <div
            class="modal-body"
          >
            <div
              class="classification-info"
            >
              <p>
                information description goes here
              </p>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button>
              Close
            </button>
          </div>
        </section>
      </div>
      <form>
        <label
          for="classques-0"
        >
          Classques 0
           
        </label>
        <select
          id="classques-0"
          name="classques-0"
          required=""
          title="Classques 0"
        >
          <option
            disabled=""
            value=""
          >
            Select an option
          </option>
        </select>
        <label
          for="classques-1"
        >
          Classques 1
           
          <a
            role="more-info-button"
          >
            <i
              class="fa"
            />
          </a>
        </label>
        <select
          id="classques-1"
          name="classques-1"
          required=""
          title="Classques 1"
        >
          <option
            disabled=""
            value=""
          >
            Select an option
          </option>
        </select>
        <label
          for="classques-2"
        >
          Classques 2
           
          <a
            role="more-info-button"
          >
            <i
              class="fa"
            />
          </a>
        </label>
        <select
          id="classques-2"
          name="classques-2"
          required=""
          title="Classques 2"
        >
          <option
            disabled=""
            value=""
          >
            Select an option
          </option>
        </select>
        <input
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Classificationpage /> functional tests Should display modal 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <form>
          <label
            for="classques-0"
          >
            Classques 0
             
          </label>
          <select
            id="classques-0"
            name="classques-0"
            required=""
            title="Classques 0"
          >
            <option
              disabled=""
              value=""
            >
              Select an option
            </option>
          </select>
          <label
            for="classques-1"
          >
            Classques 1
             
            <a
              role="more-info-button"
            >
              <i
                class="fa"
              />
            </a>
          </label>
          <select
            id="classques-1"
            name="classques-1"
            required=""
            title="Classques 1"
          >
            <option
              disabled=""
              value=""
            >
              Select an option
            </option>
          </select>
          <label
            for="classques-2"
          >
            Classques 2
             
            <a
              role="more-info-button"
            >
              <i
                class="fa"
              />
            </a>
          </label>
          <select
            id="classques-2"
            name="classques-2"
            required=""
            title="Classques 2"
          >
            <option
              disabled=""
              value=""
            >
              Select an option
            </option>
          </select>
          <input
            type="submit"
            value="Submit"
          />
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <form>
        <label
          for="classques-0"
        >
          Classques 0
           
        </label>
        <select
          id="classques-0"
          name="classques-0"
          required=""
          title="Classques 0"
        >
          <option
            disabled=""
            value=""
          >
            Select an option
          </option>
        </select>
        <label
          for="classques-1"
        >
          Classques 1
           
          <a
            role="more-info-button"
          >
            <i
              class="fa"
            />
          </a>
        </label>
        <select
          id="classques-1"
          name="classques-1"
          required=""
          title="Classques 1"
        >
          <option
            disabled=""
            value=""
          >
            Select an option
          </option>
        </select>
        <label
          for="classques-2"
        >
          Classques 2
           
          <a
            role="more-info-button"
          >
            <i
              class="fa"
            />
          </a>
        </label>
        <select
          id="classques-2"
          name="classques-2"
          required=""
          title="Classques 2"
        >
          <option
            disabled=""
            value=""
          >
            Select an option
          </option>
        </select>
        <input
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Classificationpage /> functional tests Should display modal 3`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="modal"
        >
          <section
            class="modal-window"
          >
            <div
              class="modal-header"
            >
              Classques 1
            </div>
            <div
              class="modal-body"
            >
              <div
                class="classification-info"
              >
                <p>
                  information description goes here
                </p>
              </div>
            </div>
            <div
              class="modal-footer"
            >
              <button>
                Close
              </button>
            </div>
          </section>
        </div>
        <form>
          <label
            for="classques-0"
          >
            Classques 0
             
          </label>
          <select
            id="classques-0"
            name="classques-0"
            required=""
            title="Classques 0"
          >
            <option
              disabled=""
              value=""
            >
              Select an option
            </option>
          </select>
          <label
            for="classques-1"
          >
            Classques 1
             
            <a
              role="more-info-button"
            >
              <i
                class="fa"
              />
            </a>
          </label>
          <select
            id="classques-1"
            name="classques-1"
            required=""
            title="Classques 1"
          >
            <option
              disabled=""
              value=""
            >
              Select an option
            </option>
          </select>
          <label
            for="classques-2"
          >
            Classques 2
             
            <a
              role="more-info-button"
            >
              <i
                class="fa"
              />
            </a>
          </label>
          <select
            id="classques-2"
            name="classques-2"
            required=""
            title="Classques 2"
          >
            <option
              disabled=""
              value=""
            >
              Select an option
            </option>
          </select>
          <input
            type="submit"
            value="Submit"
          />
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="modal"
      >
        <section
          class="modal-window"
        >
          <div
            class="modal-header"
          >
            Classques 1
          </div>
          <div
            class="modal-body"
          >
            <div
              class="classification-info"
            >
              <p>
                information description goes here
              </p>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button>
              Close
            </button>
          </div>
        </section>
      </div>
      <form>
        <label
          for="classques-0"
        >
          Classques 0
           
        </label>
        <select
          id="classques-0"
          name="classques-0"
          required=""
          title="Classques 0"
        >
          <option
            disabled=""
            value=""
          >
            Select an option
          </option>
        </select>
        <label
          for="classques-1"
        >
          Classques 1
           
          <a
            role="more-info-button"
          >
            <i
              class="fa"
            />
          </a>
        </label>
        <select
          id="classques-1"
          name="classques-1"
          required=""
          title="Classques 1"
        >
          <option
            disabled=""
            value=""
          >
            Select an option
          </option>
        </select>
        <label
          for="classques-2"
        >
          Classques 2
           
          <a
            role="more-info-button"
          >
            <i
              class="fa"
            />
          </a>
        </label>
        <select
          id="classques-2"
          name="classques-2"
          required=""
          title="Classques 2"
        >
          <option
            disabled=""
            value=""
          >
            Select an option
          </option>
        </select>
        <input
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
